aws/resource_aws_kinesis_firehose_delivery_stream.go

```diff
@@ -2304,34 +2398,80 @@ func resourceAwsKinesisFirehoseDeliveryStreamDelete(d *schema.ResourceData, meta
+		})
 		if err != nil {
 			if isAWSErr(err, firehose.ErrCodeResourceNotFoundException, "") {
-				return 42, "DESTROYED", nil
+				return &firehose.DeliveryStreamDescription{}, FirehoseDeliveryStreamStatusDeleted, nil
 			}
-			return nil, "failed", err
+			return nil, "", err
 		}
 
-		return resp.DeliveryStreamDescription, *resp.DeliveryStreamDescription.DeliveryStreamStatus, nil
+		return resp.DeliveryStreamDescription, aws.StringValue(resp.DeliveryStreamDescription.DeliveryStreamStatus), nil
 	}
 }
 
-func waitForKinesisFirehoseDeliveryStreamDeletion(conn *firehose.Firehose, deliveryStreamName string) error {
+func firehoseDeliveryStreamSSEStateRefreshFunc(conn *firehose.Firehose, sn string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := conn.DescribeDeliveryStream(&firehose.DescribeDeliveryStreamInput{
+			DeliveryStreamName: aws.String(sn),
+		})
+		if err != nil {
+			return nil, "", err
+		}
+
+		return resp.DeliveryStreamDescription, aws.StringValue(resp.DeliveryStreamDescription.DeliveryStreamEncryptionConfiguration.Status), nil
+	}
+}
+
+func waitForKinesisFirehoseDeliveryStreamDeletion(conn *firehose.Firehose, deliveryStreamName string, timeout time.Duration) error {
+	_, err := firehoseDeliveryStreamWaitUntilTargetState(conn, deliveryStreamName, timeout, []string{firehose.DeliveryStreamStatusDeleting}, []string{FirehoseDeliveryStreamStatusDeleted})
+
+	return err
+}
+
+func firehoseDeliveryStreamWaitUntilTargetState(conn *firehose.Firehose, deliveryStreamName string, timeout time.Duration, pending, target []string) (*firehose.DeliveryStreamDescription, error) {
 	stateConf := &resource.StateChangeConf{
-		Pending:    []string{"DELETING"},
-		Target:     []string{"DESTROYED"},
-		Refresh:    firehoseStreamStateRefreshFunc(conn, deliveryStreamName),
-		Timeout:    20 * time.Minute,
+		Pending:    pending,
+		Target:     target,
+		Refresh:    firehoseDeliveryStreamStateRefreshFunc(conn, deliveryStreamName),
+		Timeout:    timeout,
 		Delay:      10 * time.Second,
 		MinTimeout: 3 * time.Second,
 	}
+	v, err := stateConf.WaitForState()
+	if err != nil {
+		return nil, err
+	}
 
+	return v.(*firehose.DeliveryStreamDescription), nil
+}
+
+func firehoseDeliveryStreamSSEWaitUntilTargetState(conn *firehose.Firehose, deliveryStreamName string, timeout time.Duration, pending, target []string) error {
+	stateConf := &resource.StateChangeConf{
+		Pending:    pending,
+		Target:     target,
+		Refresh:    firehoseDeliveryStreamSSEStateRefreshFunc(conn, deliveryStreamName),
+		Timeout:    timeout,
+		Delay:      10 * time.Second,
+		MinTimeout: 3 * time.Second,
+	}
 	_, err := stateConf.WaitForState()
+	if err != nil {
+		return err
+	}
 
-	return err
+	return nil
+}
+
+func isKinesisFirehoseDeliveryStreamOptionDisabled(v interface{}) bool {
+	options := v.([]interface{})
+	if len(options) == 0 {
+		return true
+	}
+	e := options[0].(map[string]interface{})["enabled"]
```

Comment:
To prevent potential panics:

```suggestion
	m := options[0].(map[string]interface{})

	var enabled bool

	if v, ok := m["enabled"]; ok {
		enabled = v.(bool)
	}
	
	return !enabled
```